include/nvexec/detail/memory.cuh

```diff
@@ -110,4 +116,243 @@ namespace nvexec::STDEXEC_STREAM_DETAIL_NS {
+    }
+
+    bool do_is_equal(const std::pmr::memory_resource& other) const noexcept override {
+      return this == &other;
+    }
+  };
+
+  struct monotonic_buffer_resource : std::pmr::memory_resource {
+    static constexpr size_t block_alignment = 256;
+
+    struct block_descriptor_t {
+      void* ptr{};
+      size_t total{};
+    };
+
+    std::pmr::memory_resource* upstream;
+    std::vector<block_descriptor_t> allocated_blocks;
+
+    size_t space{};
+    void* current_ptr{};
+
+    monotonic_buffer_resource(std::size_t bytes, std::pmr::memory_resource* upstream)
+      : upstream(upstream)
+      , space(bytes) {
+      block_descriptor_t first_block{upstream->allocate(space, block_alignment), space};
+      current_ptr = first_block.ptr;
+      allocated_blocks.push_back(first_block);
+    }
+
+    ~monotonic_buffer_resource() {
+      for (block_descriptor_t& block: allocated_blocks) {
+        upstream->deallocate(block.ptr, block.total, block_alignment);
+      }
+    }
+
+    block_descriptor_t get_current_block() {
+      return allocated_blocks.back();
+    }
+
+    size_t get_next_space() {
+      return get_current_block().total * 2;
+    }
+
+    void* do_allocate(size_t bytes, size_t alignment) override {
+      void* ptr = std::align(alignment, bytes, current_ptr, space);
+
+      if (ptr == nullptr) {
+        space = std::max(bytes, get_next_space());
+        ptr = current_ptr = upstream->allocate(space, block_alignment);
+        allocated_blocks.push_back(block_descriptor_t{current_ptr, space});
+      }
+
+      current_ptr = static_cast<char*>(current_ptr) + bytes;
+      space -= bytes;
+
+      return ptr;
+    }
+
+    void do_deallocate(void* /* ptr */, size_t /* bytes */, size_t /* alignment */) override {
+    }
+
+    bool do_is_equal(const std::pmr::memory_resource& other) const noexcept override {
+      return this == &other;
+    }
+  };
+
+  struct synchronized_pool_resource : std::pmr::memory_resource {
+    constexpr static size_t alignment = 256;
+
+    struct block_descriptor_t {
+      static constexpr unsigned int min_bin = 3;
+
+      void* ptr{};
+      unsigned int bin{};
+      size_t bytes{};
+
+      block_descriptor_t(size_t bytes)
+        : bin(cuda::std::bit_width(bytes))
+        , bytes(1ull << bin) {
+        if (bin < min_bin) {
+          bin = min_bin;
+          bytes = 1ull << bin;
+        }
+      }
+
+      block_descriptor_t(void* ptr)
+        : ptr(ptr) {
+      }
+    };
+
+    struct ptr_comparator_t {
+      bool operator()(const block_descriptor_t& a, const block_descriptor_t& b) const {
+        return a.ptr < b.ptr;
+      }
+    };
+
+    struct size_comparator_t {
+      bool operator()(const block_descriptor_t& a, const block_descriptor_t& b) const {
+        return a.bytes < b.bytes;
+      }
+    };
+
+    using cached_blocks_t = std::multiset<block_descriptor_t, size_comparator_t>;
+    using busy_blocks_t = std::multiset<block_descriptor_t, ptr_comparator_t>;
```

Comment:
Why `multiset` here? There won't be two block descriptors with the same ptr, will there?